sklearn/metrics/tests/test_pairwise_distances_reduction.py

```diff
@@ -0,0 +1,362 @@
+    dtype=np.float64,
+):
+
+    rng = np.random.RandomState(seed)
+    spread = 100
+    X = rng.rand(n_samples, n_features).astype(dtype) * spread
+    Y = rng.rand(n_samples, n_features).astype(dtype) * spread
+
+    # Haversine distance only accepts 2D data
+    if metric == "haversine":
+        X = np.ascontiguousarray(X[:, :2])
+        Y = np.ascontiguousarray(Y[:, :2])
+
+    parameter = (
+        10
+        if PairwiseDistancesReduction is PairwiseDistancesArgKmin
+        # Scaling the radius slightly with the numbers of dimensions
+        else 10 ** np.log(n_features)
+    )
+
+    dist_par_X, indices_par_X = PairwiseDistancesReduction.compute(
+        X,
+        Y,
+        parameter,
+        metric=metric,
+        # Taking the first
+        metric_kwargs=_get_dummy_metric_params_list(metric, n_features)[0],
+        # To be sure to use parallelization
+        chunk_size=n_samples // 4,
+        strategy="parallel_on_X",
+        return_distance=True,
+    )
+
+    dist_par_Y, indices_par_Y = PairwiseDistancesReduction.compute(
+        X,
+        Y,
+        parameter,
+        metric=metric,
+        # Taking the first
+        metric_kwargs=_get_dummy_metric_params_list(metric, n_features)[0],
+        # To be sure to use parallelization
+        chunk_size=n_samples // 4,
+        strategy="parallel_on_Y",
+        return_distance=True,
+    )
+
+    ASSERT_RESULT[PairwiseDistancesReduction](
+        dist_par_X,
+        dist_par_Y,
+        indices_par_X,
+        indices_par_Y,
+    )
+
+
+@pytest.mark.parametrize("n_features", [50, 500])
+@pytest.mark.parametrize("translation", [10 ** i for i in [4, 8]])
+@pytest.mark.parametrize("metric", PairwiseDistancesReduction.valid_metrics())
+@pytest.mark.parametrize(
+    "PairwiseDistancesReduction",
+    [PairwiseDistancesArgKmin],
+)
+def test_euclidean_translation_invariance(
+    n_features,
+    translation,
+    metric,
+    PairwiseDistancesReduction,
+    n_samples=1000,
+    dtype=np.float64,
+):
+    # The reduction must be translation invariant.
+    parameter = (
+        10
+        if PairwiseDistancesReduction is PairwiseDistancesArgKmin
+        # Scaling the radius slightly with the numbers of dimensions
+        else 10 ** np.log(n_features)
+    )
+
+    rng = np.random.RandomState(0)
+    spread = 100
+    X = rng.rand(n_samples, n_features).astype(dtype) * spread
+    Y = rng.rand(n_samples, n_features).astype(dtype) * spread
+
+    # Haversine distance only accepts 2D data
+    if metric == "haversine":
+        X = np.ascontiguousarray(X[:, :2])
+        Y = np.ascontiguousarray(Y[:, :2])
+
+    reference_dist, reference_indices = PairwiseDistancesReduction.compute(
+        X,
+        Y,
+        parameter,
+        metric=metric,
+        metric_kwargs=_get_dummy_metric_params_list(metric, n_features)[0],
+        return_distance=True,
+    )
+
+    dist, indices = PairwiseDistancesReduction.compute(
+        X + 0,
+        Y + 0,
```

Comment:
Nice catch. I completely forgot about that. :open_mouth: 

> Instead of such a generic test that should work for all reductions, I think we could have dedicated tests for each reduction. For instance, a test that compares the argkmin reduction with a naive scipy cdist + numpy implementation for all supported metrics on data close enough to 0.0 to put ourselves in a favorable numerical setting.

That's a good point. Let's add it in another PR?

Reply:
Mainly callable functions and specialization of distance metrics present in `PAIRWISE_DISTANCE_FUNCTIONS`:

https://github.com/scikit-learn/scikit-learn/blob/d5e045fd600bf7f3edd984fe7f8599124a67f7f6/sklearn/metrics/pairwise.py#L1372-L1385

Reply:
Alright, let's focus on the nominal cases for this this test.